src/wh_nvm.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * src/wh_flash.c
+ *
+ * Simple helper library
+ *
+ */
+
+#include <stddef.h>     /* For NULL */
+#include <string.h>     /* For memset, memcpy */
+
+#include "wolfhsm/wh_common.h"
+#include "wolfhsm/wh_error.h"
+
+#include "wolfhsm/wh_nvm.h"
+
+
+int wh_Nvm_Init(whNvmContext* context, const whNvmConfig *config)
+{
+    if (    (context == NULL) ||
+            (context->cb == NULL) ) {
+        return WH_ERROR_BADARGS;
+    }
+
+    /* No callback? Return ABORTED */
+    if (context->cb->Init == NULL) {
+        return WH_ERROR_ABORTED;
+    }
+    return context->cb->Init(context->context, config);
+}
+
+int wh_Nvm_Cleanup(whNvmContext* context)
+{
+    if (    (context == NULL) ||
+            (context->cb == NULL) ) {
+        return WH_ERROR_BADARGS;
+    }
+
+    /* No callback? Return ABORTED */
+    if (context->cb->Cleanup == NULL) {
+        return WH_ERROR_ABORTED;
+    }
+    return context->cb->Cleanup(context->context);
+}
+
+int wh_Nvm_GetAvailable(whNvmContext* context,
+        uint32_t *out_avail_size, whNvmId *out_avail_objects,
+        uint32_t *out_reclaim_size, whNvmId *out_reclaim_objects)
+{
+    if (    (context == NULL) ||
+            (context->cb == NULL) ) {
+        return WH_ERROR_BADARGS;
+    }
+
+    /* No callback? Return ABORTED */
+    if (context->cb->GetAvailable == NULL) {
+        return WH_ERROR_ABORTED;
+    }
+    return context->cb->GetAvailable(context->context,
+            out_avail_size, out_avail_objects,
+            out_reclaim_size, out_reclaim_objects);
+}
+
+
+int wh_Nvm_AddObject(whNvmContext* context, whNvmMetadata *meta,
+        whNvmSize data_len, const uint8_t* data)
+{
+    if (    (context == NULL) ||
+            (context->cb == NULL) ) {
+        return WH_ERROR_BADARGS;
+    }
+
+    /* No callback? Return ABORTED */
+    if (context->cb->AddObject == NULL) {
+        return WH_ERROR_ABORTED;
+    }
+    return context->cb->AddObject(context->context, meta, data_len, data);
```

Comment:
Just at thought: If we aren't going to NULL check `data` or check `data_len` then we should make sure to note in the callback interface in `wh_nvm.h` what arguments the callbacks are responsible for checking (E.g. at the "NVM flash layer") and what will be checked by the NVM layer before dispatching the callback. As we start to add more back-ends, then documenting these things will be important.

Reply:
Invoking AddObject with NULL data and 0 data length is absolutely ok.  The lower level nvm handler should be the one verify "goodness" of parameters and they are all required to do that.  This helper library is only stopping obvious insanity.